# Unit 2 Part 2: Types

## Interactive Programming Exercises

## Learning Outcomes
- Write programs using different types

*This notebook looks at code fragments. After completing these exercises you MUST then go on to write full programs - see the programming exercises in the workbook.

Answers to exercises are given at the end*


## Why Types
Variables have types in Java. Types collect values into groups that have the same properties such as being numbers or text or a true/false value. Each is stored in a different way (using a different representation).

A strong type system where variables have types is important as it helps

- tell the compiler what values you (the programmer) intend so put in it so can catch your mistakes
    - eg if you said a variable was to hold a number then put a word in it you are confused and it can tell you
    
- allow it to allocate the right space to store the values 
    - eg a string needs more space than an integer
    
- tell the compiler what operations make sense so it can catch your mistakes
    - eg if you said a variable was to hold a string but then try to multiply with it you are confused and it can tell you

Strong types are part of defenive programming: they help prevent buggy programs.

### Exercise 1

Read the following code fragment and decide what you think it will do. Run it and see if you are right.
Explain what it does.


**Double click then write your explanation here**


In [7]:
String surname;
char initial;
int age;
double height;
boolean enrolled;

surname = "AHMED";
initial = 'Y';
age = 18;
height = 1.72;
enrolled = true;

System.out.println("The student's name is " + initial + " " + surname);
System.out.println("Their age is " + age + " and height is " + height);
System.out.println("Their enrolled state is " + enrolled);

The student's name is P BROWN
Their age is 18 and height is 1.72
Their enrolled state is true


### Exercise 2

Edit the following fragment so that it stores the equivalent details of the student:
Ellie Nadri who is 19 has height 1m 59cm tall and is not yet enrolled.

Make sure it prints the correct details,

In [7]:
String surname;
char initial;
int age;
double height;
boolean enrolled;

surname = "AHMED";
initial = 'Y';
age = 18;
height = 1.72;
enrolled = true;

System.out.println("The student's name is " + initial + " " + surname);
System.out.println("Their age is " + age + " and height is " + height);
System.out.println("Their enrolled state is " + enrolled);

The student's name is P BROWN
Their age is 18 and height is 1.72
Their enrolled state is true


## Different core types
There are a whole series of core types provided as standard as seen 

- whole numbers ie integers (int)
    - ... -3, -2, -1, 0, 1, 2, 3 ... 999, 1000, ...
    
- numbers with a decimal point ie floating point numbers (double)
    - 1.5, 3.14159, 0.1, ...
    
- characters ie the equivalent of key presses on the keyboard (char)
    - 'a', 'b', ... 'A', 'B' ... '1', ''2' ... ":', "?' ...
    
- sequences of characters ie Strings (String)
    - "Paul", "123", "abc456", "Hello there"
    
- truth values ie boolean values (boolean)
    - true, false

### Exercise 3

Edit the following fragment to instead record details of a species of beetle including its name, mass in grams (a whole number) and length in cm, storing each piece of information in a separate variable. It should record the details of the Goliath Beetle, printing the message:
```
The Goliath Beetle weighs up to 115g 
and measures up to 11.5 cm in length as larvae.
```
Use appropriate names for variables.

HINT: make sure you remember + operators between each separate thing being joined together in your print statement.

In [12]:
String surname;
char initial;
int age;
double height;
boolean enrolled;

surname = "AHMED";
initial = 'Y';
age = 18;
height = 1.72;
enrolled = true;

System.out.println("The student's name is " + initial + ". " + surname);
System.out.println("Their age is " + age + " and height is " + height + "m tall");
System.out.println("Their enrolled state is " + enrolled);

The student's name is Y. AHMED
Their age is 18 and height is 1.72m tall
Their enrolled state is true


## Strings and concatenation

A string is made up of a sequence of characters. The main operation you can do on Strings is to 'concatenate' two strings (ie stick them together end to end). You have been doing this already in print statements. You use the + operator, but note when used with strings it is concatenation not addition.

### Exercise 4
Complete the following code so that it stores "abra" in variable magic1 and "cadabra" in variable magic2. The concatenation of the two should be stored in the third variable magic_word. Finally magic_word should be printed.

In [ ]:
String magic1;
String magic2;
String magic_word;

// WRITE THREE ASIGNMENTS HERE TO CONSTRUCT THE MAGIC WORD TO PRINT

System.out.println(magic_word);

### Exercise 5
Predict what the following code will print. Then run it to see if you were right. Explain what happens.


**Double click then write your explanation here**


In [21]:
String s1 = "123";
String s2 = "321";
String answer;

answer = s1 + s2;

System.out.println(answer);

123321


## Integers
Integers are whole numbers. You can do arithmetic on them (```+ - * /```).

### Exercise 6
Predict what the following code will print. Then run it to see if you were right. Explain what happens.

In [22]:
int s1 = 123;
int s2 = 321;
int answer;

answer = s1 + s2;

System.out.println(answer);

444


## Converting Strings to Integers

Strings are just sequences of characters so the string "123" is stored as '1' followed by '2' followed by '3' just as it looks on the page. It isn't a number anymore than the string "wombat" is.

The number 123 is actually stored in a special (binary) format that the computer can easily do arithmetic on.

When strings like "123" and integers like 123 are printed they look identical. 

A built-in method Integer.parseInt takes strings and converts them to numbers (if they do convert to a number).

### Exercise 7

Run the following then work out what it does line by line. Then add a line at the end
```java
System.out.println(text_input + 1);
```
What do you expect it to do?

In [28]:
Scanner scanner = new Scanner(System.in);

String text_input;
int n;
int plus1;

System.out.println("Input a number");
text_input = scanner.nextLine();           // Note you input text (keyboard strokes) - a series of characters
n = Integer.parseInt(text_input);  // So need to convert it to an integer


System.out.println("You typed " + text_input);
System.out.println("Which is the integer " + n);
plus1 = n+1;
System.out.println("You can do arithmetic on an (not on a string) integer so n plus 1 is " + plus1);

Input a number
9
You typed 9
Which is the integer 9
You can do arithmetic on an (not on a string) integer so n plus 1 is 10
91


### Exercise 8

A quick way to add up all the numbers from 1 to any positive number n is using the formula n(n+1)/2. eg 
```1 + 2 + 3 = 3x4/2 = 6```
```1 + 2 + 3 + 4 + 5 + 6 = 6x7/2 = 21```

The following code fragment is intended to do that but gets the answers WRONG. It contains a bug. Read and run the code. Work out what it is doing wrong and correct the code.

In [18]:
Scanner scanner = new Scanner(System.in);

String text_input;
int n;
int answer;

System.out.println("Input a number");
text_input = scanner.nextLine();           // Note you input text (keyboard strokes) - a series of characters
n = Integer.parseInt(text_input);  // So need to convert it to aninteger

answer = n * (n - 1) / 2;

System.out.println("The sum to " + n + " is " + answer);


Input a number
6
The sum to 6 is 15


## Decimal numbers - Doubles

Integers are whole numbers. What if you want more accurate numbers? You then need decimals. One way is the type double. You can do arithmetic on decimal numbers though the details is different (remember in primary school you had to learn how to do it - how do you deal with the decimal points?). Decimal numbers are called floating point numbers because of the way they are stored - the decimal point floats around depending on the number rather than always being in the same place.

### Exercise 9

Predict exactly what the following will print then run the code.

In [31]:
double d1 = 1.5;
double d2 = 1.5;
double answer;

answer = d1 + d2;

System.out.println(answer);

3.0


### Exercise 10
When you take the average of whole numbers you may not get a whole number answer...

Run the following code then see if you can explain what happens

In [37]:
int d1 = 3;
int d2 = 2;
double answer1;
double answer2;

answer1 = (d1 + d2) / 2.0;
answer2 = (d1 + d2) / 2;

System.out.println(answer1);
System.out.println(answer2);

2.5
2.5


## Strong type systems

The above shows that dealing doubles is complicated. It is really easy to write code that does the wrong calculation. The compiler rather than helping spot errors does invisible conversions. The type system has been compromised to 'help' programmers by giving 'convenient' short cuts. The problem is it does not know for sure what you were trying to do.

A better langiuage design would make you convert values explicitly. If you tried to divide an integer by a double it should raise a compile-time error and stop you.

Strong type systems help avoid buggy programs! Bugs like this about the precise type of number have cost billions of dollars (eg causing rockets carrying satellites to crash/explode). Java has quite a strong type system but it is not strong enough! It lets mistakes through that it could catch.

Even when a language gives you shortcuts it is better to always explicitly include the conversion code so it is clear to other readers what you intend.

## Inputting decimals
As with ints, doubles have to be first input as a string and then converted. The methods Double.parseDouble does this in an equivalent way to Integer.parseInt. **Parsing** which these methods do is just the operation of converting a series of characters into someother format.

### Exercise 11
Convert the following to allow decimals to be input and work completely with double arithmetic rather than integer arithmetic and so give accurate answers.

In [41]:
Scanner scanner = new Scanner(System.in);

String text_input;
int n1;
int n2;
int answer;

System.out.println("Give me a number");
text_input = scanner.nextLine();           // Note you input text (keyboard strokes) - a series of characters
n1 = Integer.parseInt(text_input);  // So need to convert it to aninteger

System.out.println("Give me a second number");
text_input = scanner.nextLine();           // Note you input text (keyboard strokes) - a series of characters
n2 = Integer.parseInt(text_input);  // So need to convert it to aninteger


answer = (n1 + n2) / 2;

System.out.println("The average is " + answer);


Give me a number
2
Give me a second number
3
The average is 2


*Once you have done the above exercises (and understand how the concepts work) move on to doing the actual programming exercises from the handbook, writing full programs. You must be able to write full programs, not just fragments.*

## Solutions

### Exercise 1

The code prints the details (name, age height and enrolled state) of a student.

It first declares a series of variables, each of a different appropriate type to store the separate pieces of information about a student. The surname is a sequence of characters so a String. The initial is a single character so of type, char. The age is a whole number so int. The height is a decimal number so cannot be stored as an int and so is instead a double. Whether they are enrolled is either true or false so a boolean.

Assignment is used to store the appropriate value in each. Then each variable is used at the appropriate point in a print statement to print out an appropriate message.

### Exercise 2

```java
String surname;
char initial;
int age;
double height;
boolean enrolled;

surname = "NADRI";
initial = 'E';
age = 19;
height = 1.59;
enrolled = false;

System.out.println("The student's name is " + initial + ". " + surname);
System.out.println("Their age is " + age + " and height is " + height + "m tall");
System.out.println("Their enrolled state is " + enrolled);
```

### Exercise 3
```java
String beetle_name;
int max_weight;
double max_length;

beetle_name = "Goliath Beetle";
max_weight = 115;
max_length = 11.5;
enrolled = true;

System.out.println("The " + beetle_name + " weighs up to " + max_weight + "g");
System.out.println("and measures up to " + max_length + "cm in length as larvae.");
```

### Exercise 4
```java
String magic1;
String magic2;
String magic_word;

magic1 = "abra";
magic2 = "cadabra";
magic_word = magic1+magic2;
System.out.println(magic_word);
```
### Exercise 5

It prints 123321

As both values are strings NOT numbers, + here means concatenation NOT addition so it sticks the two strings of digits end to end.

### Exercise 6
It prints 444

This time both variables s1 and s2 are integers. When two variables that are declared as integers are added using +, the two values will be added arithmetically.

### Exercise 7
If you say type in 9 when asked then the code first prints it as a string then as an integer, having converted it and stored it in the int variable n. Both look the same when printed. However, you can do artihmetic on the int version such as n + 1. So if 9 was input ti will print 10

If you add a line to print 
```
text_input + 1
```
it does not do addition, it does concatenation as it is working with a String and that is the only thing that makes sense for a string. 


### Exercise 8

A minus sign was mistakenly used instead of a plus sign. The calculation should be
```java
answer = n * (n + 1) / 2;
```

```java
Scanner scanner = new Scanner(System.in);

String text_input;
int n;
int answer;

System.out.println("Input a number");
text_input = scanner.nextLine();           // Note you input text (keyboard strokes) - a series of characters
n = Integer.parseInt(text_input);  // So need to convert it to aninteger

answer = n * (n + 1) / 2;

System.out.println("The sum to " + n + " is " + answer);
```

### Exercise 9
It prints 3.0 (not 3) 3 is an integer whereas 3.0 is a double. As we did floating point addition on doubles we get a double answer. ALways add .0 to whole numbers if you want them to be treated as floating point numbers.

### Exercise 10

Both d1 and d2 are integers. We add them as integers to get an integer total. So if we divide by an integer we get an integer answer as we do integer division giving the answer 2 (as integers round down).
```
5 / 2 = 2
```
On the other hand  when we divide by a double the more accurate double division is done. The 5 is automatically turned into a double so that double division can be done
```
5.0 / 2.0 = 2.5
```

If we change everything to doubles then accurate floating point arithmetic is done either way even when dividing by an integer. Even when the 2 is an integer, it is converted to the more accurate double version so double arithmetic can be done. 

The program this time was 
```java
double d1 = 3.0;
double d2 = 2.0;
double answer1;
double answer2;

answer1 = (d1 + d2) / 2.0;
answer2 = (d1 + d2) / 2;

System.out.println(answer1);
System.out.println(answer2);
```
Both answers are 2.5

### Exercise 11
You need to change all the int types to double, change the integer conversions to double conversions and change int literal values (here 2) to double literal values (2.0) to be absolutely clear you intend accurate double arithmetic throughout.

```java
Scanner scanner = new Scanner(System.in);

String text_input;
double n1;
double n2;
double answer;

System.out.println("Give me a number");
text_input = scanner.nextLine();           // Note you input text (keyboard strokes) - a series of characters
n1 = Double.parseDouble(text_input);  // So need to convert it to aninteger

System.out.println("Give me a second number");
text_input = scanner.nextLine();           // Note you input text (keyboard strokes) - a series of characters
n2 = Double.parseDouble(text_input);  // So need to convert it to aninteger


answer = (n1 + n2) / 2.0;

System.out.println("The average is " + answer);
```
